### ISSUE TITLE
Proposition: Vite plugin

### ISSUE THREAD

**Issue by Odonno** [amiceli/vitest-cucumber#145]
The proposition is to go beyond the current provided CLI and to offer a built-in Vite plugin to generate the js/ts file architecture from feature files, relatively similar to the [TanStack Router vite plugin](https://tanstack.com/router/latest/docs/framework/react/quick-start#configure-the-vite-plugin).

## Implementation details

The idea is to provide a fairly straightforward solution that automatically generate feature/scenario/step definitions files and the underlying folder architecture. But file generation is only the starting point as the ideal solution would be to keep the architecture completely in sync on every feature file change:

* new feature = new file generated, prefilled with incomplete scenario tests
* new scenario/step = addition of new lines of code (+updates of import if necessary)
* removed step/scenario = removal of existing lines of code
* removed feature = removal of definition file
* renamed file = renaming of definition file
* renamed scenario/step = renaming of the existing scenario/step in the definition file
* [Bonus] reusing the same step = refactoring by importing/exporting the step definition from a new generated file
* [Alter bonus] un-reusing the same step = remove the generated step definition file & rollback changes on the only file using this step
* [Bonus] detect unsaved (and invalid?) definition files = avoid syncing and restart syncing once all definition files are saved

Having a predictable file architecture would allow such complex workflow letting the developer purely focus on the changes to make in each step definition (and possibly in-between scenarii and steps).

### Setup & configuration

A minimal vite config could look like this:

```ts
import { defineConfig } from 'vite';
import { CucumberVite } from 'cucumber-vite';

export default defineConfig({
  plugins: [
    CucumberVite(),
    // your preferred framework plugin here
  ],
});
```

The previous example will use the default configuration. A fully configurated vite config using the `CucumberVite` plugin could look like this:

```ts
import { defineConfig } from 'vite';
import { CucumberVite } from 'cucumber-vite';

export default defineConfig({
  plugins: [
    CucumberVite({
        enabled: false, // it is possible to completely disable syncing, if we need to make a big refactor change for example
        mode: "feature",
        featuresFolder: "./src/__features__",
        definitionsFolder: "./src/__features__", // feature & definitions files can be colocated or not in the same folder
        sharedDefinitionsFolder: "./src/__features__/shared", // for the refactoring part when a same step is used at least twice
    }),
    // your preferred framework plugin here
  ],
});
```

#### Modes

The goal is to craft the ideal solution which would be the full autopilot/full sync mode. However, it can be 1. easier to build in steps and it can be 2. interesting to offer a way to provide a progressive enhancement from nothing to the full syncing mode, for cases when integrating this plugin into a brownfield project is impossible or when the full-sync mode feels too "aggressive".

A set of 3 modes can be integrated and built upon in the solution. From simplest to more feature complete:

1. `feature` level, only generate new scenario on file change
2. `scenario` level, like 1. but it can also generate new steps in a scenario definition when new steps are added on file change
3. `full sync`, the one depicted earlier which offers the best experience possible and syncing on every change

Question: quid of the renaming and the removal of file/scenario/step on `feature` and `scenario` levels?

### `new feature`

Let's take the following Gherkin definition from an hypothetic feature file:

```gherkin
Feature: Improve my unit tests
  Scenario: Use vitest-cucumber in my unit tests
    Given Developer using feature file
    And   Using vitest-cucumber
    When  I run my unit tests
    Then  I know if I forgot a scenario
```

This can generate a new definition file like this:

```ts
import { loadFeature, describeFeature } from '@amiceli/vitest-cucumber';
import { expect } from 'vitest';

const feature = await loadFeature('./file.feature');

describeFeature(feature, ({ Scenario }) => {
  Scenario('Use vitest-cucumber in my unit tests', ({ Given, When, Then, And }) => {
    Given('Developer using feature file', (ctx) => {
      ctx.todo();
    });
    And('sing vitest-cucumber', (ctx) => {
      ctx.todo();
    });
    When('I run my unit tests', (ctx) => {
      ctx.todo();
    });
    Then('I know if I forgot a scenario', (ctx) => {
      ctx.todo();
    });
  });
});
```

### `refactoring`

The refactoring is applied when and only when the same step is used at least twice, both the step text and the step definition code. This refactoring can happen whether a feature file reuse the same step across 2+ scenarii or whether a step already exist in a scenario of another feature file.

Let's start with a feature file that will trigger a refactoring:

```gherkin
Feature: Improve my unit tests
  Scenario: Use vitest-cucumber in my unit tests
    Given Developer using feature file
    And   Using vitest-cucumber
    When  I run my unit tests
    Then  I know if I forgot a scenario

  Scenario: Generate definition files using the ViteCucumber plugin
    Given Developer using feature file
    And   Using ViteCucumber plugin
    When  I create a new feature file
    Then  A new definition file is generated
```

This will generate both the definition file and the shared step definition file.

```ts
import { loadFeature, describeFeature } from '@amiceli/vitest-cucumber';
import { expect } from 'vitest';
import { givenDeveloperUsingFeatureFile } from './shared/givenDeveloperUsingFeatureFile';

const feature = await loadFeature('./file.feature');

describeFeature(feature, ({ Scenario }) => {
  Scenario('Use vitest-cucumber in my unit tests', ({ Given, When, Then, And }) => {
    givenDeveloperUsingFeatureFile(Given);
    And('sing vitest-cucumber', (ctx) => {
      ctx.todo();
    });
    When('I run my unit tests', (ctx) => {
      ctx.todo();
    });
    Then('I know if I forgot a scenario', (ctx) => {
      ctx.todo();
    });
  });

  // the second scenario will be here...
});
```

Note: the generated "shared step" will be named `givenDeveloperUsingFeatureFile.ts`

```ts
export const givenDeveloperUsingFeatureFile = (Given) => {
  Given('Developer using feature file', (ctx) => {
    ctx.todo();
  });
};
```

Note 2: we can imagine that the shared step naming convention can be configured globally via the configuration (either via defined presets `"camelCase" | "pascalCase" | "kebabCase" | "snakeCase"` and/or simply via a function `(stepName: string) => string`)

The decision of having a single step per file is based on previous experiences where having N+ shared steps in a single step would re-trigger more tests than expected on a single step change because the test runner is incapable to detect which step has been changed and so which import is triggering a change.

Again, by enabling the autopilot mode, this opinionated decision is a no-brainer as the "shared step" file will be created/removed automatically by the Vite cucumber plugin.

**Comment by flowt-au** [comment]
Thanks for all your great proposals. With regards to this one, an addition that would be great is to add a config option for a  code fragment that pulls in custom setup code. I have a support file and some basic assignments that are at the top of each of my tests. It would be great to define all that in a file to include and to add that code fragment via the config filepath when generating a new test file.
eg:
```javascript
// All my custom setup code
import { graphDB, msRequest, msBroadcast, msReceive, loadFeature, describeFeature, setupServerForTests, checkValidReceivedMessage, isISO8601, toBoolean } from 'test/vitest/__tests__/helpers/startupTestBusinessLogicClasses.js'

const loopServerIn = inject('loopServerIn')
import { basetestdata } from 'test/vitest/__tests__/testData/baseData.js'

let forceLoopServerIn

const useServerForTest = (forceLoopServerIn !== null) ? forceLoopServerIn : loopServerIn
```
Thanks again for continuing the development of this great library.
Murray

**Comment by amiceli** [comment]
Vite plugin is really a good idea but I will see that in another project like `vite-cucumber-plugin`.

It will use `vitest-cucumber` for CLI etc but all code to detect file changes and update tests/shared according changes, generate new tests etc, I don't see that in same project.

I never code a vite plugin but I can check how it works, docs, examples etc and maybe scratch an `alpha` project.

**Comment by flowt-au** [comment]
Good luck! :-)

**Comment by amiceli** [comment]
For first alpha version, I planned these scenarii.
Want do you think ?

~~~gherkin
Feature: vitest-cucumber plugin

    Background:
        Given My "*.feature" files are in "src/__tests__/"
        And   My "*.spec.ts" files are in "src/__tests__/"

    Scenario: Create spec file for new feature file
        Given "src/__tests__/awesome.spec.ts" doesn't exists
        When  I write "src/__tests__/awesome.feautre"
        Then  vitest-cucumber create "src/__tests__/awesome.spec.ts"

    Scenario: Add scenario to spec file
        Given "src/__tests__/awesome.spec.ts" hasn't scenario
        When  I add a scenario into "src/__tests__/awesome.feautre"
        """
        Scenario: example
            Given I'm new scenario
            Then  I'm added into spec file
        """
        Then  vitest-cucumber add new scenario in "src/__tests__/awesome.spec.ts"

    Scenario: Remove scenario in feature file
        Given "src/__tests__/awesome.feautre" has "example" scenario
        When  I remove "example" scenario in feature file
        Then  vitest-cucumber remove "example" scenario in "src/__tests__/awesome.spec.ts"
~~~

**Comment by flowt-au** [comment]
Looks like a simple enough starting point. I look forward to testing when you have something. :-)

**Comment by amiceli** [comment]
I published an alpha version `4.1.1-alpha.1`.
Currently vitest plugin add/remove scenario according feature file.

Now I'm working for steps but if a step is removed from feature we can remove it or comment it.
What do you think ?

**Comment by flowt-au** [comment]
I support commenting rather than deleting. If one has written the assertions for a given step, it would be good to keep them in case you need to re-implement the assertions logic in the new step, or somewhere else. If one doesn't need the step it is trivial to delete it oneself.

**Comment by Odonno** [comment]
I would prefer deletion, and using the git commit history to retrieve previous steps if possible.

If there is a debate on this, we can think about adding a configuration parameter such as `onDeleteAction: 'delete' | 'comment' | 'none'`.

**Comment by flowt-au** [comment]
Ok. I think the suggestion of the `onDeleteAction` config parameter is a good idea. People have different use-cases and work flows. :-)

**Comment by amiceli** [comment]
Yes it's a good idea.
I plan it in vitest plugin. I updated PR with features list.

**Comment by amiceli** [comment]
I published a new alpha version `4.1.1-alpha.2`.
Now plugin handle add/remove Background, Scenario, Rule and Steps.

Take care currently remove, really remove code, I'm working on `onDeleteAction ` ;).

**Comment by amiceli** [comment]
I publish a new version `4.1.1-alpha.3`.
I added `onDeleteAction ` with `delete` or `comment`.
Plugin also update arrow function arguments.

You can test it, asap I will publish a stable version ;).

**Comment by amiceli** [comment]
I published a beta version `4.1.1-beta.1`.
To debrief plugin:
- create / update spec file
- comment or remove  Scenario, Rule, Step, Background etc
- Allow to run a command to format code like `npx biome check --write` ;)

Asap I will publish a stable version, we will open new issue for fix.

I will check later for step definition, it's a little hard to understand.
For example I've same steps in Backgrounds for plugin tests: [example](https://github.com/amiceli/vitest-cucumber/blob/feat/vitest-plugin/src/plugin/__tests__/scenario/scenario-ast.spec.ts).

But I need variable declared in parent in Background step.

**Comment by amiceli** [comment]
I close, can be re-open if needed or open new issue.
